Store management: adds interface Store as a dependency of quiz instead of plain QAs

Loveuse opened the pull request:
Store exposes the methods Extract, Setup, GetQAs, QAsLen and it is part
of a new file, store.go.
Moves QA definition near to the Store interface declaration.
Previous CSV extractor is a store that retrieves QAs from a reader
based on the CSV format.

Loveuse committed:
Update the extractor package and QAs structures, Extract and Run methods sign

Modify the name of the files in the extractor package to be more appropriate.
Question and answer are stored together in a single object and a slice of it is
used to store QAs.
Extract method reflects the changes applied to QAs, as well as the quiz struct.
Run uses a timer to show the available time to a user in order to complete a quiz.

Loveuse committed:
Added interface Store as dependency of quiz instead of plain QAs

Store exposes the methods Extract, Setup, GetQAs, QAsLen and it is part
of a new file store.go
Moves QA definition near to the Store interface declaration.
Previous CSV extractor is a store that retrieves QAs from a reader
based on the CSV format.

Loveuse committed:
Remove exercise1 binary

indiependente commented on exercise1/main.go:
```diff
@@ -23,21 +28,22 @@ func main() {
-	qaExtractor := fromcsv.Extractor{}
-	questions, answers, err := qaExtractor.Extract(filePtr)
+	store := &csv.Store{}
+	QAs, err := store.Extract(filePtr)
```
Why do you need QAs in the main package if you have the Store as dependency in the Quiz?

indiependente commented on exercise1/quiz/extractor/csv/csv_store.go:
```diff
@@ -0,0 +1,56 @@
+	"io"
+	"log"
+
+	store "github.com/Loveuse/gophercises/exercise1/quiz/extractor"
```
why importing extractor package as store?

indiependente commented on exercise1/quiz/extractor/csv/csv_store.go:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	if len(QAs) < 1 {
+		return nil, errors.New("could not read any question")
```
declare known errors as exported variables in this package

indiependente commented on exercise1/quiz/extractor/csv/csv_store.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Setup assigns the questions and answers for the store
+func (s *Store) Setup(qas []store.QA) {
```
swap this with a constructor

indiependente commented on exercise1/quiz/extractor/store.go:
```diff
@@ -0,0 +1,38 @@
+// - Setup: loads the QAs in the store
+// - GetQAs: returns QAs
+// - LenQAs: returns the number of QAs
+type Store interface {
```
this is probably not a store

indiependente commented on exercise1/quiz/quiz.go:
```diff
@@ -3,56 +3,46 @@ package quiz
+		fmt.Print(qa.Question + ": ")
 		if _, err := fmt.Scanf("%s\n", &answer); err != nil {
-			return fmt.Errorf("could not read the answer for %s: ", question)
+			return fmt.Errorf("could not read the answer for %s: ", qa.Question)
```
use errors

Loveuse committed:
clean Store interface and separate the extraction of QAs

- extract package renamed
- clean the store interface having only methods Get and Set
- move QA definition to store/qa package
- QA Store uses custom json.Marshal and json.Unmarshal for QA
  type to keep answers unexported
- add the method LenQAs to QA store in order to get the number of QAs
- define errors through github.com/pkg/errors

Loveuse committed:
Merge branch 'store-mgmt' of github.com:Loveuse/gophercises into store-mgmt

Loveuse committed:
add comments to exported errors in QA store
